### ISSUE TITLE
The conan recipe is horribly outdated

### ISSUE THREAD

**Issue by LunarWatcher** [machinezone/IXWebSocket#96]
<sub>The reason I'm opening this issue here is because the link to the conan recipe is [in the readme](https://github.com/machinezone/IXWebSocket#build). The recipe has several issues that, judging by recent commits as well, appear to be by design, or as a result of not being officially updated</sub>

I've been wanting to try out this library, but because of incompatible build systems, I have to use Conan. 

I couldn't find IXWebSocket in any conan repository - only a recipe. So I went ahead and cloned it, and ran the command I'll be using for all these: `conan create . IXWebSocket/stable`. Output on the first go:

```
ERROR: Failed requirement 'OpenSSL/1.1.1b@zinnion/stable' from 'IXWebSocket/1.5.7@IXWebSocket/stable'
ERROR: Unable to find 'OpenSSL/1.1.1b@zinnion/stable' in remotes
```

I took a look at the dependencies, and they appear to be practically identical to the ones found in the official conan repo. So I replaced the dependencies with the same versions from `conan/stable`. Before that, I noticed the conan recipe is actually using a fork off this repo that's over 500 commits behind, and it's running 1.5.7, which is non-existent for this repo.

Anyway, re-run, fails:

```
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
/usr/bin/clang++  -D_GLIBCXX_USE_CXX11_ABI=1 -I/root/.conan/data/IXWebSocket/1.5.7/IXWebSocket/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/include -I/root/.conan/data/OpenSSL/1.1.1b/conan/stable/package/e93595837cd71d31cd792a7899549d9f13bcd0cc/include -I/root/.conan/data/zlib/1.2.11/conan/stable/package/6d87124bcf4cc4435e13c44634d61b77404bef31/include  -m64 -stdlib=libstdc++ -O3 -DNDEBUG    -std=gnu++1z -o CMakeFiles/test_package.dir/test_package.cpp.o -c /programming/conanOrig/test_package/test_package.cpp
[100%] Linking CXX executable bin/test_package
/usr/bin/cmake -E cmake_link_script CMakeFiles/test_package.dir/link.txt --verbose=1
/usr/bin/clang++    -m64 -stdlib=libstdc++ -O3 -DNDEBUG        -rdynamic CMakeFiles/test_package.dir/test_package.cpp.o  -o bin/test_package  -L/root/.conan/data/IXWebSocket/1.5.7/IXWebSocket/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/lib  -L/root/.conan/data/OpenSSL/1.1.1b/conan/stable/package/e93595837cd71d31cd792a7899549d9f13bcd0cc/lib  -L/root/.conan/data/zlib/1.2.11/conan/stable/package/6d87124bcf4cc4435e13c44634d61b77404bef31/lib -Wl,-rpath,/root/.conan/data/IXWebSocket/1.5.7/IXWebSocket/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/lib:/root/.conan/data/OpenSSL/1.1.1b/conan/stable/package/e93595837cd71d31cd792a7899549d9f13bcd0cc/lib:/root/.conan/data/zlib/1.2.11/conan/stable/package/6d87124bcf4cc4435e13c44634d61b77404bef31/lib -lixwebsocket -lssl -lcrypto -ldl -lpthread -lz
/root/.conan/data/IXWebSocket/1.5.7/IXWebSocket/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/lib/libixwebsocket.a(IXSocketFactory.cpp.o): In function `ix::createSocket(bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
IXSocketFactory.cpp:(.text+0x52): undefined reference to `ix::SocketOpenSSL::SocketOpenSSL(int)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
CMakeFiles/test_package.dir/build.make:97: recipe for target 'bin/test_package' failed
make[2]: *** [bin/test_package] Error 1
make[2]: Leaving directory '/programming/conanOrig/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6'
CMakeFiles/Makefile2:70: recipe for target 'CMakeFiles/test_package.dir/all' failed
make[1]: *** [CMakeFiles/test_package.dir/all] Error 2
make[1]: Leaving directory '/programming/conanOrig/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6'
Makefile:86: recipe for target 'all' failed
make: *** [all] Error 2
ERROR: IXWebSocket/1.5.7@IXWebSocket/stable (test package): Error in build() method, line 15
        cmake.build()
        ConanException: Error 512 while executing cmake --build '/programming/conanOrig/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6' '--' '-j1'
```

When I was setting this up, I didn't notice the version. So I replaced the homepage URL without replacing the version, which resulted in another exception when it tried to get the archive, because 1.5.7 doesn't exist in this repo. I bumped the version as well, re-build, and:

```
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
/usr/bin/clang++  -D_GLIBCXX_USE_CXX11_ABI=1 -I/root/.conan/data/IXWebSocket/5.0.0/IXWebSockets/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/include -I/root/.conan/data/OpenSSL/1.1.1b/conan/stable/package/e93595837cd71d31cd792a7899549d9f13bcd0cc/include -I/root/.conan/data/zlib/1.2.11/conan/stable/package/6d87124bcf4cc4435e13c44634d61b77404bef31/include  -m64 -stdlib=libstdc++ -O3 -DNDEBUG    -std=gnu++1z -o CMakeFiles/test_package.dir/test_package.cpp.o -c /programming/conan-IXWebSocket/test_package/test_package.cpp
/programming/conan-IXWebSocket/test_package/test_package.cpp:23:30: error: no member named 'WebSocket_MessageType_Message' in namespace 'ix'
      if (messageType == ix::WebSocket_MessageType_Message) {
                         ~~~~^
/programming/conan-IXWebSocket/test_package/test_package.cpp:24:14: error: no member named 'cout' in namespace 'std'
        std::cout << str << std::endl;
        ~~~~~^
/programming/conan-IXWebSocket/test_package/test_package.cpp:24:34: error: no member named 'endl' in namespace 'std'
        std::cout << str << std::endl;
                            ~~~~~^
/programming/conan-IXWebSocket/test_package/test_package.cpp:17:5: error: no viable conversion from '(lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5)' to 'const ix::OnMessageCallback'
      (aka 'const function<void (const shared_ptr<ix::WebSocketMessage> &)>')
    [](ix::WebSocketMessageType messageType,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:421:7: note: candidate constructor not viable: no known conversion from '(lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5)' to
      'std::nullptr_t' (aka 'nullptr_t') for 1st argument
      function(nullptr_t) noexcept
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:432:7: note: candidate constructor not viable: no known conversion from '(lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5)' to
      'const std::function<void (const std::shared_ptr<ix::WebSocketMessage> &)> &' for 1st argument
      function(const function& __x);
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:441:7: note: candidate constructor not viable: no known conversion from '(lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5)' to
      'std::function<void (const std::shared_ptr<ix::WebSocketMessage> &)> &&' for 1st argument
      function(function&& __x) noexcept : _Function_base()
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/std_function.h:465:2: note: candidate template ignored: substitution failure
      [with _Functor = (lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5), $1 = void]: no type named 'type' in 'std::result_of<(lambda at /programming/conan-IXWebSocket/test_package/test_package.cpp:17:5) &(const
      std::shared_ptr<ix::WebSocketMessage> &)>'
        function(_Functor);
        ^
/programming/conan-IXWebSocket/test_package/test_package.cpp:17:5: note: candidate function
    [](ix::WebSocketMessageType messageType,
    ^
/root/.conan/data/IXWebSocket/5.0.0/IXWebSockets/stable/package/2d89e03e00941cc4186fea9b793094a560d9fb7b/include/ixwebsocket/IXWebSocket.h:80:60: note: passing argument to parameter 'callback' here
        void setOnMessageCallback(const OnMessageCallback& callback);
                                                           ^
4 errors generated.
CMakeFiles/test_package.dir/build.make:65: recipe for target 'CMakeFiles/test_package.dir/test_package.cpp.o' failed
make[2]: *** [CMakeFiles/test_package.dir/test_package.cpp.o] Error 1
make[2]: Leaving directory '/programming/conan-IXWebSocket/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6'
CMakeFiles/Makefile2:70: recipe for target 'CMakeFiles/test_package.dir/all' failed
make[1]: *** [CMakeFiles/test_package.dir/all] Error 2
make[1]: Leaving directory '/programming/conan-IXWebSocket/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6'
Makefile:86: recipe for target 'all' failed
make: *** [all] Error 2
ERROR: IXWebSocket/5.0.0@IXWebSockets/stable (test package): Error in build() method, line 15
        cmake.build()
        ConanException: Error 512 while executing cmake --build '/programming/conan-IXWebSocket/test_package/build/47edf16f7c6208a569a5a85106a6734e405b9db6' '--' '-j1'
```

I also actually can't find some of these fields in the fork. The new variant (`WebSocketMessageType`, as outlined in the readme of this project) at least exists in this repo. The [file in question](https://github.com/machinezone/IXWebSocket/blob/d4115880b9038704d82d4e7b3b562a3c3309c015/ixwebsocket/IXWebSocketMessageType.h) also only exists here - I can't find either of these in the repo the recipe points to.

Also, off the top of my head, [this will probably fail](https://github.com/Zinnion/conan-IXWebSocket/blob/master/test_package/test_package.cpp) (if that's an actual concept - I'm not that familiar with developing Conan recipes myself), because not everyone happens to have a websocket server at `localhost:8080` running. 

So for the hell of it, I used my fixed package to test. I used the example using localhost:8080 on the main site, with the imports:
```
#include <iostream>
#include <ixwebsocket/IXWebSocket.h>
#include <ixwebsocket/IXSocket.h>
#include <ixwebsocket/IXWebSocketMessageType.h>
```

as well as the conan package I built and a basic CMakeLists.txt I pretty much [borrowed from the Conan docs](https://docs.conan.io/en/latest/getting_started.html). It compiled with no errors. 

But to get there, I had to change the repo, bump the version, and remove the tests (while updating the test could've worked as well, I'm not familiar with the library, or creating conan recipes, so I didn't want to push my luck). 

Could the conan recipe be updated? Pushing it to conan-center as well could be a good idea.

**Comment by bsergean** [comment]
Hi @LunarWatcher. Thanks for all the efforts trying to get this building with conan. I've actually been playing with vcpkg and there's a 'port' for it that got merged recently.

This port I linked too is un-maintained and ancient indeed. The README for this project has examples, and the unittest is a good place to look for up to date code.

I am *very* unfamiliar with conan, so if you want to make a PR for it I'll accept it. I don't know much about the steps for pushing something conan-center.

**Comment by LunarWatcher** [comment]
I haven't tried publishing anything to conan-center, but I already have an account on Bintray, so it shouldn't be too hard. 

When you say you want a PR, do you want the recipe in this repo? Most of the other conan recipes I've seen use separate repos. It's no problem either way, I just need to make sure I understand what you want. (A documentation update PR is kinda implied here either way)

Either way, I'll need to set up some tests, but I *think* once I've done that, I should be able to upload it. Now that there are actual tests, I might be able to run it from there instead of manually creating tests. Some packages create their own tests, others (like Boost), appear to use [placeholder C++ files](https://github.com/conan-community/conan-boost/blob/release/1.70.0/test_package/test.cpp) linking to the internal tests. Not really sure - like I said, I don't have much experience with creating conan recipes. 

Like I said, the current one works minimally, but IIRC, there needs to be tests in order to upload it to the central repo. And since you said you're not familiar with conan, conan center is [the central conan repository](https://bintray.com/conan/conan-center). Packages can be added there freely, but there's constraints and I think review before that happens, but I haven't tried publishing a package. Uploading it to Bintray alone allows the repo to be added as a remote, and it enables the recipe/any built binaries for compatible platforms to be downloaded without manually calling `conan create` to install it anywhere necessary. 

Users can also upload recipes for third party libraries, so it's not limited to the owners. The specific guidelines for both are listed [here](https://docs.conan.io/en/latest/uploading_packages/bintray/conan_center_guide.html#conan-center-flow) if you're interested.

**Comment by bsergean** [comment]
Ok I'll take a look. Documentation change is probably the only thing needed then. I've been thinking about adding a small 'Build' wiki page (or markdown or whatever is the most convenient) where we could explain all the different ways to build this lib.

**Comment by LunarWatcher** [comment]
Update on the issue:

I've re-written the entire recipe from scratch. It's in [the same repo](https://github.com/lunarwatcher/conan-ixwebsocket), but it's no longer a fork because, like I said, re-writes. 

I added a test as well, but it currently excludes Windows due to an issue that may or may not be of importance:

```
Unable to connect to echo.websocket.org on port 80, error: Either the application has not called WSAStartup, or WSAStartup failed
```

This is Windows-only from the looks of it (no-repro on an Ubuntu 18.04 Docker container). I have practically no experience with this library - could I be missing some library linking here, or is Windows just broken?

Anyway, beyond that, the recipe should be ready. I ended up creating a test using [websocket.org](https://www.websocket.org/echo.html) as an echo server, which works fine on Linux (aside two messages and a race condition, but that's not really a big deal as long as it's at lesat returning something).

**Comment by bsergean** [comment]
Excellent !

ix::initNetSystem();

needs to be called early on Windows (see test_runner.cpp)

I'm curious about the race conditions.

**Comment by LunarWatcher** [comment]
I'll push the new code in a second - that call fixed it on Windows. As for the race condition, that appears to have been a loop problem. Weirdly though, it doesn't happen on Windows, only Linux (might be a container issue of course, but I kinda had that coming for using a while loop to dispatch messages :] ). 

In this I noticed I created a bug with linking the vendored version of MbedTLS on Linux. I'll fix that tomorrow. Beyond that, the test seems to be fine, and the library is doing its job. I'll leave another comment once I've patched up the vendored MbedTLS issue on Linux, but after that, I think the recipe should be up to date. It's still untested on Apple (which I can't test because I don't have access to an Apple device) and Android, but I imagine Android is pretty close to Linux.

**Comment by LunarWatcher** [comment]
I've fixed the build issue in Linux with mbedTLS (it was a linking order issue - added a safeguard to make sure it's linked in the right order). 

The recipe should be mostly ready now. It could of course do with more testing, but I can't break it locally on two different operating systems, using three different compilers. Like I said yesterday, I can't test Apple, but it should be easy to patch any issues there. 

I'll see if I can figure out how to push it to Conan Center, and any issues with the recipe can be handled as they show up (issues are also enabled in the recipe repo). I can submit a documentation update once I get it published if you want.

**Comment by LunarWatcher** [comment]
I should probably update this. I figured I'd wait until I heard back from Bintray, but I haven't heard anything. 

I've finished off the rest of the inclusion requirements and submitted it for inclusion. I haven't gotten a reply yet. When it gets included, just adding the dependency is enough. For the record, it's:

```
IXWebSocket/5.0.0@LunarWatcher/stable
```

In the meanwhile, it can be accessed by adding a new remote (`conan remote add remote_name_here https://api.bintray.com/conan/oliviazoe0/conan-packages`). 

I'm not sure if you want to make a temporary documentation update while the request gets processed, but either way, the recipe is up.

**Comment by bsergean** [comment]
Thanks for the update Olivia. I'm in vacation now so it's ok if this drags
a little.

I'll try to build and setup conan myself, maybe next week. I have a mac so
I can test on this platform.

**Comment by bsergean** [comment]
I have updated the main README with new instructions. Tell me if it doesn't make sense. I still haven't have a chance to build this with conan myself ...

**Comment by LunarWatcher** [comment]
I had to set up CI as a part of the inclusion criteria, and the Mac builds pass there at least. But that's also a full build, and not direct local inclusion. That being said, it works fine locally on Linux and Windows (no local build - pulling straight from Bintray). Still waiting for a reply from Bintray (I really have no idea how long this will take at this point). 

The instructions appear to be fine, but you have a copy-paste error on the remote line (an accidental `).` snuck in at the end of the code block).

**Comment by bsergean** [comment]
Cool thanks. I just fixed the typo.

**Comment by bsergean** [comment]
I just released 5.0.4 which had a build fix on Windows, if this matter.
Not sure what we should do with this ticket, can we close it?
Did bintray ever replied ?

**Comment by LunarWatcher** [comment]
I haven't gotten a reply yet. No idea when I can expect it either. Thanks for the message though - I'll set up some builds and push 'em (I see I've missed a few versions - I need to find a decent notification system). 

I suppose it makes sense to close this issue. I'll rather leave a comment on this if I get a reply from them - the problem itself is fixed.

**Comment by bsergean** [comment]
Great thanks.

I'm trying to document what's changing in here -> https://github.com/machinezone/IXWebSocket/blob/master/CHANGELOG.md

There wasn't too much since 5.0.0, no changes to the core library, but there was a windows fix I just made, caused by something I introduced ... at some point :)

**Comment by LunarWatcher** [comment]
It seems the former Bintray inclusion process [has been deprecated](https://docs.conan.io/en/latest/uploading_packages/bintray/conan_center_guide.html), so I guess I'll never get a reply. 

I'll poke around and see what I can find on inclusion. I'll have to check if the project qualifies, and how badly it breaks the recipe. The existing conan repo and everything I've mentioned earlier should:tm: still work regardless of inclusion. It seems Bintray is used as well, so it shouldn't break that at least, regardless of the inclusion outcome. I'll keep you updated when I figure out the main stuff.

**Comment by bsergean** [comment]
Cool good to know.

**Comment by LunarWatcher** [comment]
Sorry about the delay. 

From what I can tell, it's just a different build process. The teaks should be minor and I'm already working on that. There doesn't seem to be any requirements. All I need is (in theory:tm:) a PR and to be in the early access program, which I requested and was let into a few days ago. It's not specifically exclusive, it's actually very open. Anyway, I'm gonna be working on the recipe and testing over the next few days. There's a few factors I'm not sure about yet, such as how to go about mitigating the few builds that're somehow broken - for an instance, the release with Windows bugs a while ago, and how libraries specify the version of C++ they're using. Abeit some C++11 libraries doesn't specify anything, so it might be fine. 

Anyway, I'll send you a link whenever I submit the PR

**Comment by LunarWatcher** [comment]
[PR submitted](https://github.com/conan-io/conan-center-index/pull/677). There might be more work to do before it gets merged, but I'll stay on top of that as problems show up

**Comment by bsergean** [comment]
Great thanks.

**Comment by LunarWatcher** [comment]
Looks like everything is ready. Waiting on a final review (ended up bumping the version, but the version before that was cleared), so it should be merged some time tonight. I'll submit a documentation change PR (here) for conan usage whenever my PR for conan center gets merged in (Conan 1.21.x includes several changes, which I need to get into before I submit a documentation PR). There's a heavy reduction in the amount of versions available, but it's much easier to keep up to date now. It'll also be easier to access (no more explicit remotes to use the library)

**Comment by bsergean** [comment]
Very cool. Thanks for all the packaging work.
